src/Marten/V4/EventStoreSpike.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections.Generic;
+
+namespace Marten.V4
+{
+    public class MyMetadata
+    {
+        public Guid CorrelationId { get; set; }
+        public string UserId { get; set; }
+    }
+
+    public enum StreamState
+    {
+        CreateNew,
+        Append,
+        AppendOrCreate,
+        Pending, // Not sure this is necessary
+        History
+    }
+
+    // Similar to Refit. Marten would whip up an actual implementation
+    // for this by inheriting from the base DocumentSession, and adding
+    // implementations for the event collection.
+    public interface IMyDocumentSession : IDocumentSession
+    {
+        EventCollection<string, object> Events { get; }
+
+        // Could be more than one Event collection here as an equivalent
+        // to stream type.
+    }
+
+    // Thinking the document session would smuggle the metadata
+    // to the event collection at SaveChanges() time
+
+    // You *could* bake in optimistic concurrency per stream here
+
+    // *These types* would be codegenerated ala Refit
```

Comment:
As we discussed on Gitter, that needs more detailed discussion and consideration, as it's a longer topic and move the discussion on this idea to separate issue. We should do more research on how other frameworks handle metadata. We can also schedule some call to discuss that.